core/services/ocr2/plugins/ocr2keeper/evm21/upkeepstate/store.go

```diff
@@ -104,16 +131,34 @@ func (u *upkeepStateStore) Close() error {
 // SelectByWorkIDs returns the current state of the upkeep for the provided ids.
 // If an id is not found, the state is returned as StateUnknown.
 // We first check the cache, and if any ids are missing, we fetch them from the scanner.
-// TODO: fetch from DB
 func (u *upkeepStateStore) SelectByWorkIDsInRange(ctx context.Context, start, end int64, workIDs ...string) ([]ocr2keepers.UpkeepState, error) {
 	states, ok := u.selectFromCache(workIDs...)
 	if ok {
 		// all ids were found in the cache
 		return states, nil
 	}
+
+	// fetch values from chain to populate the cache with missing values
 	if err := u.fetchPerformed(ctx, start, end); err != nil {
 		return nil, err
 	}
+
+	idsWithUnknownState := []string{}
```

Comment:
You can have this functionality within `fetchDromDB`

Reply:
In what way? At this point we know both the ids and the states. Passing the states to `fetchFromDB` along with the ids seems like doing the same thing in another place.